syntax = "proto2";

package GstDebugger;

enum Action {
	ADD = 0;
	REMOVE = 1;
}

message Value {
	required uint64 gtype = 1;
	optional int32 internal_type = 2;
	required bytes data = 3;
}

message PropertyInfo {
	required string name = 1;
	required string nick = 2;
	required string blurb = 3;
	required int32 flags = 4;
	required Value default_value = 5;
}

message PropertyValue {
	required string name = 1;
	required string object = 2;
	required Value value = 3;
}

message PropertyRequest {
	optional string name = 1;
	required string object = 2;
}

message MessageRequest {
	required Action action = 1;
	required int32 type = 2;
}

message MessageInfo {
	required int32 type = 1;
	required uint64 timestamp = 2;
	required string source = 3;
	required uint32 seqnum = 4;
	required bytes structure_data = 5;
}

message LogRequest {
	required Action action = 1;
	required int32 level = 2;
	optional string category = 3;
}

message QueryWatchRequest {
	required int32 type = 1;
}

message BufferWatchRequest {
	required bool send_data = 1;
}

message EventWatchRequest {
	required int32 type = 1;
}

message PadWatchRequest {
	required Action action = 1;
	optional string pad = 2;
	
	oneof pad_watch_type {
		QueryWatchRequest query = 3;
		BufferWatchRequest buffer = 4;
		EventWatchRequest event = 5;
	}
}

message EventInfo {
	required int32 type = 1;
	required uint64 timestamp = 2;
	required uint32 seqnum = 3;
	required string pad = 4;
	required bytes structure_data = 5;
}

message QueryInfo {
	required int32 type = 1;
	required string pad = 2;
	required bytes structure_data = 3;
}

message TypeDescriptionRequest {
	enum Type {
		FACTORY = 0;
		ENUM_FLAGS = 1;
		KLASS = 2;
	}
	
	required string name = 1;
	required Type type = 2;
}

message Command {
	oneof command_type {
		TypeDescriptionRequest type_description = 1;
		bool debug_categories_list = 2;
		MessageRequest message = 3;
		LogRequest log = 4;
		PadWatchRequest pad_watch = 5;
		string log_threshold = 6;
		bool entire_topology = 7;
		PropertyRequest property = 8;
	}
}

message DebugCategories {
	repeated string category = 1;
}

message LogInfo {
	required int32 level = 1;
	required string category = 2;
	required string file = 3;
	required string function = 4;
	required int32 line = 5;
	required string object = 6;
	required string message = 7;
}

message EnumFlagsValue {
	required string name = 1;
	required int32 value = 2;
	required string nick = 3;
}

message EnumFlagsType {
	enum EnumFlagsKind {
		ENUM = 0;
		FLAGS = 1;
	}

	required string type_name = 1;
	required EnumFlagsKind kind = 2;
	repeated EnumFlagsValue values = 3;
}

message FactoryMeta {
	required string key = 1;
	required string value = 2;
}

message PadTemplate {
	required string name_template = 1;
	required int32 direction = 2;
	required int32 presence = 3;
	required string caps = 4;
}

message FactoryType {
	required string name = 1;
	repeated FactoryMeta metadata = 2;
	repeated PadTemplate templates = 3;
}

message ElementKlass {
	required string name = 1;
	repeated PropertyInfo property_info = 2;
}

message TopologyLink {
	required string src_pad = 1;
	required string sink_pad = 2;
}

message TopologyElement {
	required string path = 1;
	required string type_name = 2;
	required bool is_bin = 3;
	required string factory_name = 4;
}

message TopologyPad {
	required string path = 1;
	required bool is_ghostpad = 2;
	required int32 direction = 3;
	required int32 presence = 4;
	optional PadTemplate template = 5;
}

message TopologyInfo {
	required Action action = 1;

	oneof topology_type {
		TopologyLink link = 2;
		TopologyElement element = 3;
		TopologyPad pad = 4;
	}
}

message GStreamerData {
	oneof info_type {
		DebugCategories debug_categories = 1;
		LogInfo log_info = 2;
		EnumFlagsType enum_flags_type = 3;
		FactoryType factory = 4;
		MessageInfo message_info = 5;
		Command confirmation = 6;
		EventInfo event_info = 7;
		TopologyInfo topology_info = 8;
		QueryInfo query_info = 9;
		PropertyValue property_value = 10;
		ElementKlass element_klass = 11;
	}
}
